fem/src/Lua.F90

```diff
@@ -0,0 +1,386 @@
+  end subroutine
+
+end interface ! }}}
+
+abstract interface ! luafun {{{
+function luafun(L) result(n)
+  import
+  type(c_ptr), value :: L
+  integer(kind=c_int) :: n
+end function
+end interface ! }}}
+!-------------------------------------------------------------------------------
+
+CONTAINS
+
+FUNCTION lua_getusertable(L, name) result(t)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: name(*)
+  real(kind=c_double), pointer :: t(:)
+
+  type(c_ptr) :: cp_raw, cp_data
+  integer(kind=c_int) :: len
+  t => NULL()
+
+  call lua_getfield(L%L, LUA_GLOBALSINDEX, name)
+  cp_raw = lua_touserdata(L%L,-1)
+  call get_userdataptr(L%L, cp_raw, cp_data, len) ! this should typecase cp_raw to "NumArray*" and extract size and len out of it
+  call c_f_pointer(cp_data, t, shape=[len])
+end function
+
+function lua_tolstring(L, n, slen) result(sp)
+  type(c_ptr) :: L
+  integer(kind=c_int) :: n
+  character(kind=c_char, len=:), pointer :: sp
+
+  character(kind=c_char, len=:), allocatable :: s
+  type(c_ptr) :: c_s
+  integer(kind=c_int) :: slen
+
+  c_s = lua_tolstring_c(L, n, slen)
+  call c_f_pointer(c_s, sp)
+end function
+
+function luaL_checkstring(L, n, slen) result(sp)
+  type(c_ptr) :: L
+  integer(kind=c_int) :: n
+  character(kind=c_char, len=:), pointer :: sp
+
+  character(kind=c_char, len=:), allocatable :: s
+  type(c_ptr) :: c_s
+  integer(kind=c_int) :: slen
+
+  c_s = luaL_checkstring_c(L, n, slen)
+  call c_f_pointer(c_s, sp)
+end function
+
+subroutine lua_runfile(L, fname)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: fname(*)
+  call lua_runfile_c(L%L, fname)
+end subroutine
+
+function lua_dostring(L, s, m) result(n)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: s(*)
+  integer(kind=c_int) :: n, m_, load_error, pcall_error
+  integer(kind=c_int), optional :: m
+  if(.not. present(m)) then
+    m_ = -1
+  else
+    m_ = m
+  end if
+  load_error = luaL_loadstring(L%L, s)
+  call check_error(L, load_error)
+  if(load_error == 0) then
+    pcall_error = lua_pcall(L%L, 0, m_, 0)
+    call check_error(L, pcall_error)
+  end if
+  n = IOR(load_error, pcall_error)
+end function
+
+real(kind=c_double) function lua_getnumber(L, s)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: s(*)
+  call lua_getfield(L%L, LUA_GLOBALSINDEX, s)
+  lua_getnumber = lua_popnumber(L)
+end function
+
+subroutine lua_poptensor(L, t)
+  type(LuaState_t) :: L
+  real(kind=c_double), intent(out) :: t(:,:)
+
+  integer :: n1, n2, i, j
+  n1 = size(t, 1)
+  n2 = size(t, 2)
+  do i = n1,1, -1
+    do j = n2,1,-1
+      t(i,j) = lua_popnumber(L)
+    end do
+  end do
+end subroutine
+
+subroutine lua_eval_f(L, fname, X, y)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: fname(*)
+  real(kind=c_double), intent(in) :: X(:)
+  real(kind=c_double), intent(inout) :: Y(:)
+
+  integer :: nx, ny, i, lstat
+  nx = size(X,1)
+  ny = size(Y,1)
+  CALL lua_getfield(L%L, LUA_GLOBALSINDEX, fname)
+  do i = 1,nx
+    CALL lua_pushnumber(L%L, X(i))
```

Comment:
With ifort, the compilation fails here, in the `lua_pushnumber` interface first argument is 
`type(c_funptr), value :: L` whereas `L % L` is `type(c_ptr)`.